README.md
# Janus CLI

Janus CLI is a Toolbelt framework that provides a simple plugin system to allow you to build plugins for your build process. 

It is important to note that Janus CLI is *not* in itself an executable but a tool to help you build your own. NPM have a [great blog post](http://blog.npmjs.org/post/118810260230/building-a-simple-command-line-tool-with-npm) that explains how to build an executable using NPM.

# Existing plugins

* [Task Runner Plugin](https://github.com/thebeansgroup/januscli-tasks)
  * As it sounds, the task runner will synchronously run tasks from a YAML file in the current repo. Useful for pre-release tasks.
* [Release Plugin](https://github.com/thebeansgroup/januscli-release)
  * This plugin handles semantic versioning, auto changelings and github releases.

# Usage

## Writing your executable file 

Our company Janus executable is open over at [thebeansgroup/tbg-januscli](https://github.com/thebeansgroup/tbg-januscli) and acts as a template for rolling out your own.

This example shows how you can use the above plugins using `ES6`:

```javascript
import janus from 'januscli';
import release from 'januscli-release';
import tasks from 'januscli-tasks';

janus.loadPlugin(release);
janus.loadPlugin(tasks);

janus.start();
```

or using `common.js` modules:

```javascript
var janus = require('januscli').default;
var release = require('januscli-release').default;
var tasks = require('januscli-tasks').default;

janus.loadPlugin(release);
janus.loadPlugin(tasks);

janus.start();
```

Once you’ve built your executable — and assuming you have named it `janus` — then you can run the tasks like any CLI. 

```zsh
$ my-repo> janus -h

  Usage: janus [options] [command]


  Commands:

    release [level_override]  Create a release of current app
    tasks                     Run tasks in project

  Options:

    -h, —help      output usage information
    -V, —version   output the version number
    -v, —verbose   Verbose mode for debugging

```

# Writing plugins

A plugin is just an extension of the main [plugin class](https://github.com/thebeansgroup/januscli/blob/master/lib/plugin.js). Within your plugin you need only do three things:
* Set CLI commands
* Set CLI options
* Add listeners and handlers for commands.

The following is a simple example of a plugin taken from the [Janus Plugin Example repo](https://github.com/thebeansgroup/januscli-plugin-example).

```javascript
import { Plugin } from 'januscli';


/**
 * Example plugin definition
 */

class Example extends Plugin {

  constructor(janus) {
    super(janus);
    this.foo = 'bar';
  }

 /**
  * name() set plugin name
  *
  */

  name() {
    return 'example';
  }


 /**
  * Event handlers
  *
  */

  events() {
    this.janus.on( this.startEventName(), this.startExampleCommand.bind(this) )
    this.janus.on( 'example2:start', this.startExample2Command.bind(this) )
  }

 /**
  * CLI options for plugin to
  * respond to.
  *
  */

  cliCommands() {
    return [
      [
        'example', // command line option name
        'Run example command', // Command line description
        this.name() // event name
      ],
      [
        'example2', // command line option name
        'Run example 2 command', // command line description
        'example2' // event name
      ]
    ]
  }

 /**
  * CLI options for plugin to
  * respond to.
  *
  * This is a decorator for https://github.com/tj/commander.js/#option-parsing
  * 
  */

  cliOptions() {
    return [
      [
        '-f, --foo <value>', // command line flags
        'Set Foo', // command line description
        'bar', // default value
        (value) => { // Callback (this differs to commander.js)
          this.foo = value;
        }
      ]
    ]
  }

  /**
  * Example command
  *
  */

  startExampleCommand() {
    this.janus.success('Example:', 'Example command started');
    this.janus.info('Example:', `Foo = ${this.foo}` );
  }

  /**
  * Example 2 command
  *
  */

  startExample2Command() {
    this.janus.success('Example:', 'Example 2 command started');
    this.janus.info('Example:', `Foo = ${this.foo}` );
  }


}

export default Example;
```

This plugin can then be consumed by your toolbelt:

```javascript
import janus from 'januscli';
import Example from './example.js';

janus.loadPlugin(Example);

janus.start();
```

Example output:

```zsh
[master][~/src/janustest] ./bin/index.js -h

  Usage: index [options] [command]


  Commands:

    example    Run example command
    example2   Run example 2 command

  Options:

    -h, --help         output usage information
    -V, --version      output the version number
    -v, --verbose      Verbose mode for debugging
    -f, --foo <value>  Set Foo
    
 [master][~/src/janustest] ./bin/index.js example
✓ Success: Example: Example command started
ℹ Info: Example: Foo = bar

[master][~/src/janustest] ./bin/index.js example2
✓ Success: Example: Example 2 command started
ℹ Info: Example: Foo = bar

[master][~/src/janustest] ./bin/index.js example2 --foo buzz
✓ Success: Example: Example 2 command started
ℹ Info: Example: Foo = buzz
```


CHANGELOG.md
# [2.0.0] - 2016-08-03
- Feature/config - [PR Link](https://github.com/thebeansgroup/januscli/pull/23)
# [1.0.0] - 2016-03-23
- Janus becomes a framework - [PR Link](https://github.com/thebeansgroup/januscli/pull/14)
 - Janus should be a framework not a CLI in itself - [Fixes #13](https://github.com/thebeansgroup/januscli/issues/13)


lib/cli.js
import commander from 'commander';
import version from './version.js';

/**
 * Global CLI setup
 */

class CLI {

  constructor(janus) {
    this.janus = janus;
    commander.version(version);
  }

 /**
  * addOption() use commander option
  * as a callback system for plugins
  *
  * @param {String} flags
  * @param {String} description
  * @param {Mixed} defaultValue
  * @param {Function} fn
  * @return {this} for chaining
  */

  addOption(flags, description, defaultValue, fn) {
    commander.option(flags, description, fn, defaultValue);
    return this;
  }

 /**
  * addCommand() use commander command
  * as a callback system for plugins
  *
  * @param {String} flags
  * @param {String} description
  * @param {String} name
  * @return {this} for chaining
  */

  addCommand(cmd, description, name) {
    commander.command(cmd)
      .description(description)
      .action(
        (arg) => this.janus.emit(name + ':start', arg)
      );
    return this;
  }

  /**
  * parse() Call the commander parse method
  *
  */

  parse() {
    commander.parse(process.argv);
  }
}

export default CLI;


lib/janus.js
import EventEmitter from 'events';
import CLI from './cli.js';
import inquirer from 'inquirer';
import Plugin from './plugin.js';
import colors from 'colors';

const plugins = {};

/**
 * Janus CLI Core Class
 */

class Janus extends EventEmitter {

  constructor() {
    super();
    this.cli = new CLI(this);
    this.inquirer = inquirer;
  }

  /**
  * loadPlugin() loads Janus plugin
  * and passes class scope
  *
  * @param {Class} plugin Plugin ES6 Class
  */

  loadPlugin(plugin) {
    const inst = new plugin(this);
    plugins[inst.name()] = inst;
  }

  /**
  * hadPlugin() test if plugin is loaded
  *
  * @param {String} name Plugin name
  */

  hasPlugin(name) {
    return !!plugins[name];
  }


  /**
  * start() Starts the lifecycle of Janus
  *
  */

  start() {
    this.cli.parse();
  }

 /**
  * Custom Janus Errors
  *
  */

  error(pre, msg, exit) {
    exit = (exit === undefined) ? true : exit;
    const error_txt = String.fromCharCode(0x2718) + " Error: ";
    console.log(error_txt.red + pre.grey + ' ' + msg);
    if(exit) process.exit(1);
  }

 /**
  * Custom Janus Success
  *
  */

  success(pre, msg) {
    const success_txt = String.fromCharCode(0x2713) + " Success: ";
    console.log(success_txt.green + pre.grey + ' ' + msg);
  }

 /**
  * Custom Janus Info Message
  *
  */

  info(pre, msg) {
    const info_txt = String.fromCharCode(0x2139) + " Info: ";
    console.log(info_txt.grey + pre.grey + ' ' + msg);
  }

}

export { Plugin };
export default new Janus();


lib/config.js
import nconf from 'nconf';
import nconfYAML from 'nconf-yaml';
import fs from 'fs';

const file = '.janus';

/**
 * Load file into config
 */

class Config {

  constructor(janus) {
    this.janus = janus;
    this.config = nconf;

    if (!this.isSetup()) {
      this.reportMissingFile();
    }
    this.readConfigFile();
  }

  isSetup() {
    let setup = true;
    try {
      fs.readFileSync(`./${file}`).toString('utf8');
    } catch (e) {
      setup = false;
    }

    return setup;
  }

  readConfigFile() {
    this.config.file({
      file,
      format: nconfYAML,
    });
  }

  reportMissingFile() {
    this.janus.error(
      'Config:', 
      "File '.janus' is missing. Please add the file and run again.",
      true
    );
  }

  get(key) {
    return nconf.get(key);
  }

  set(key, value) {
    return nconf.set(key, value);
  }

}


export default Config;


lib/plugin.js
/**
 * Plugin base class
 */

class Plugin {
  constructor(janus) {
    this.janus = janus;
    this.setupCLI();
    this.events();
  }

  /**
  * name() Default plgin method for name.
  * if method is not set in comsuming class
  * it will throw an error.
  *
  */

  name() {
    this.janus.error('CLI:', 'Plugin name not set');
  }


  /**
  * Default plgin method for event handling.
  * if method is not set in comsuming class
  * it will throw an error.
  *
  */

  events() {
    this.janus.error('CLI:', 'Plugin events not set');
  }

  /**
  * Start event name constant
  *
  */

  startEventName() {
    return this.name() + ':start';
  }

  /**
  * Finish event name constant
  *
  */

  completeEventName() {
    return this.name() + ':complete';
  }
  
 
  /**
  * setupCLI() finds plugin cli options and
  * parse.
  *
  */

  setupCLI() {
    const options = this.cliOptions();
    if(options) {
      options.forEach( this.addCLIOption.bind(this) );
    }

   const commands = this.cliCommands();
   if(commands) {
      commands.forEach( this.addCLICommand.bind(this) );
    }
  }

  /**
  * Default methods to prevent errors
  *
  */

  cliOptions() { return []; }
  cliCommands() { return []; }

  /**
  * addCLIOption() call plugin cli option 
  * on janus cli method.
  *
  */

  addCLIOption(option) {
    this.janus.cli.addOption.apply(
      this.janus.cli,
      option
    )
  }

  /**
  * addCLICommand() call plugin cli command
  * on janus cli method.
  *
  */

  addCLICommand(command) {
    this.janus.cli.addCommand.apply(
      this.janus.cli,
      command
    )
  }

}

export default Plugin;


lib/index.js
import EventEmitter from 'events';
import CLI from './cli.js';
import inquirer from 'inquirer';
import Plugin from './plugin.js';
import Config from './config.js';
import colors from 'colors';

const plugins = {};

/**
 * Janus CLI Core Class
 */

class Janus extends EventEmitter {

  constructor() {
    super();
    this.cli = new CLI(this);
    this.config = new Config(this);
    this.allowVerboseDebugMode();
    this.inquirer = inquirer;
  }

  /**
  * loadPlugin() loads Janus plugin
  * and passes class scope
  *
  * @param {Class} plugin Plugin ES6 Class
  */

  loadPlugin(plugin) {
    const inst = new plugin(this);
    plugins[inst.name()] = inst;
  }

  /**
  * hadPlugin() test if plugin is loaded
  *
  * @param {String} name Plugin name
  */

  hasPlugin(name) {
    return !!plugins[name];
  }

  allowVerboseDebugMode() {
    this.debug = false;
    this.cli.addOption(
      '-v, --verbose',
      'Verbose mode for debugging',
      null,
      () => {
        this.info('Janus:', 'Verbose mode for debugging active.');
        this.debug = true; 
      }
    )
  }

  /**
  * start() Starts the lifecycle of Janus
  *
  */

  start() {
    this.cli.parse();
  }

 /**
  * Custom Janus Errors
  *
  */

  error(pre, msg, exit) {
    exit = (exit === undefined) ? true : exit;
    const error_txt = String.fromCharCode(0x2718) + " Error: ";
    console.log(error_txt.red + pre.grey + ' ' + msg);
    if(exit) process.exit(1);
  }

 /**
  * Custom Janus Success
  *
  */

  success(pre, msg) {
    const success_txt = String.fromCharCode(0x2713) + " Success: ";
    console.log(success_txt.green + pre.grey + ' ' + msg);
  }

 /**
  * Custom Janus Info Message
  *
  */

  info(pre, msg) {
    const info_txt = String.fromCharCode(0x2139) + " Info: ";
    console.log(info_txt.grey + pre.grey + ' ' + msg);
  }

}

export { Plugin };
export default new Janus();
